Add content for container template (#1954)
pulumi/pulumi-hugo: themes/default/content/templates/container-service/aws/index.md

---

> title: Container Service on AWS

Reviewer: I am just noticing a discrepancy between this and the serverless one. This one says "Container Service on AWS" and the serverless one says "AWS Serverless Application". I think I prefer {Architecture} on {Cloud} format. So we should update the Serverless one to conform

Author: the reason the serverless one says "AWS Serverless" is because we are trying to rank on the term "aws serverless" which was identified as an opportunity by 97th floor

> title: Deploy an ECS Container Service on AWS

Reviewer 2: ```suggestion
title: Container Service on AWS
```

Author: the title format deviates here because we are trying to SEO rank for the term "deploy ecs"

> meta_desc: The Containers template makes it easy to deploy a containerized application on AWS with Pulumi and Amazon ECS.

Reviewer: I would like @cnunciato to weigh in here. Is the important thing here deploying the containerized app or is the important thing here that you are deploying the Container Service which is running a container app?

Reviewer 2: It's the second thing. Many users will be interested in deploying containerized applications, and you can extend this template to deploy an application, but I think if we center on "service" and then discuss applications we'll be good.

> meta_desc: The Containers template makes it easy to deploy an ECS containerized application on AWS with Pulumi and Amazon ECS.

Reviewer 2: ```suggestion
meta_desc: The Container Service template makes it easy to deploy a containerized application on AWS with Pulumi and Amazon ECS.
```

Reviewer: But the thing here really isn't the containizerized app right? The thing here is it deploys a container service with an load balancer (and it just so happens that there is an application running atop)

meta_image: meta.png

> card_desc: Deploy an ECS container service on AWS with Pulumi and Amazon ECS.

Reviewer 2: ```suggestion
card_desc: Deploy a container service on AWS with Pulumi and Amazon ECS.
```

layout: template
template:
    prefix: container-aws
cloud:
  name: Amazon Web Services
  slug: aws
---

> The Container Service template deploys a container service to AWS with Pulumi. You can then use the container service to build your own containerized application. The template deploys [Amazon Elastic Container Service (ECS)]({{< relref "/registry/packages/aws/api-docs/ecs/cluster" >}}) for cluster management, [AWS Fargate]({{< relref "/registry/packages/awsx/api-docs/ecs/fargateservice" >}}) to run the cluster on serverless compute, and an [Application Load Balancer]({{< relref "/registry/packages/awsx/api-docs/lb" >}}) that serves the container endpoint to the internet. It also creates an [Amazon Elastic Container Repository (ECR)]({{< relref "/registry/packages/awsx/api-docs/ecr/repository" >}}) that stores the container image. The template generates a complete Pulumi program that provisions the cloud resources and installs Nginx in a container, providing you with a working project out of the box that you can customize easily and extend to suit your needs.

Reviewer: ```suggestion
The Container Service template creates an infrastructure as code project in your favorite language that deploys a container service to AWS. You can then use the container service to build your own containerized application. The architecture includes [Amazon Elastic Container Service (ECS)]({{< relref "/registry/packages/aws/api-docs/ecs/cluster" >}}) for cluster management, [AWS Fargate]({{< relref "/registry/packages/awsx/api-docs/ecs/fargateservice" >}}) to run the cluster on serverless compute, and an [Application Load Balancer]({{< relref "/registry/packages/awsx/api-docs/lb" >}}) that serves the container endpoint to the internet. It also uses an [Amazon Elastic Container Repository (ECR)]({{< relref "/registry/packages/awsx/api-docs/ecr/repository" >}}) that stores the container image. The template generates a complete Pulumi program that provisions the cloud resources and installs Nginx in a container, providing you with a working project out of the box that you can customize easily and extend to suit your needs.
```

![An architecture diagram of the Pulumi AWS Container Service template](./architecture.png)

## Using this template

> To use this template to deploy ECS and your own containerized application, make sure you've [installed Pulumi]({{< relref "/docs/get-started/install" >}}) and [configured your AWS credentials]({{< relref "/registry/packages/aws/installation-configuration#credentials" >}}), then create a new [project]({{< relref "/docs/intro/concepts/project" >}}) using the template in your language of choice:

Author: since we are trying to rank for the term "deploy ecs", i changed your suggestions a little bit

### Using your own container image

If you already have a container image you'd like to build your application with, you can do so either by replacing placeholder content in the `app` folder or by configuring the stack to point to another folder on your computer with the `image` setting:

```bash
$ pulumi config set path ../my-existing-image/build
$ pulumi up
```

### Cleaning up

You can cleanly destroy the stack and all of its infrastructure with [`pulumi destroy`]({{< relref "/docs/reference/cli/pulumi_destroy" >}}):

```bash
$ pulumi destroy
```

## Learn more

Congratulations! You're now well on your way to managing a production-grade containerzed application on AWS with Pulumi --- and there's lots more you can do from here:

* Discover more architecture templates as they're available in [Templates &rarr;]({{< relref "/templates" >}})
* Dive into the AWS package by exploring the [API docs in the Registry &rarr;]({{< relref "/registry/packages/aws" >}})

> * Expand your understanding of how Pulumi works in [Architecture &amp; Concepts &rarr;]({{< relref "/docs/intro/concepts" >}})

Reviewer: Same feedback as the serverless one.

> * Read up on the latest new features [in the Pulumi Blog &rarr;]({{< relref "/blog" >}})

Reviewer: I would actually link it to the specifically the containers tag within the blog if you can

The template requires no additional configuration. Once the new project is created, you can deploy it immediately with [`pulumi up`]({{< relref "/docs/reference/cli/pulumi_up" >}}):

```bash
$ pulumi up
```

When the deployment completes, Pulumi exports the following [stack output]({{< relref "/docs/intro/concepts/stack#outputs" >}}) values:

url
: The HTTP URL for the container's endpoint.

Output values like these are useful in many ways, most commonly as inputs for other stacks or related cloud resources. The computed `url`, for example, can be used from the command line to open the newly deployed container service in your favorite web browser:

```bash
$ open $(pulumi stack output url)
```

## Customizing the project

Projects created with the Container Service template expose the following [configuration]({{< relref "/docs/intro/concepts/config" >}}) settings:

container_port
: Specifies the port mapping for the container. Defaults to port 80.

cpu
: Specifies the amount of CPU to use with each task or each container within a task. Defaults to 512.

memory
: Specifies the amount of memory to use with each task or each container within a task. Defaults to 128.

image

> : The container image that is used to build the application. Defaults to the dockerfile in the `app` folder.

Reviewer: This technically isn't the container image right? It is the Dockerfile?

Author: @cnunciato can you advise on the right wording here?

Author: just updated it here, lmk if that works: https://github.com/pulumi/pulumi-hugo/pull/1954/commits/6c5ef01af8514b22633a1287c8517108ae4ea4b5

> : Specifies the location of the dockerfile used to build the container image that is run. Defaults to the dockerfile in the `app` folder.

Reviewer: ```suggestion
: Specifies the location of the Dockerfile used to build the container image that is run. Defaults to the Dockerfile in the `app` folder.
```

If you already have a container image you'd like to build your container service with, you can do so either by replacing the Dockerfile in the `app` folder or by configuring the stack to point to another folder on your computer with the `image` setting:

```bash
$ pulumi config set path ../my-existing-image/build
$ pulumi up
```

### Cleaning up

You can cleanly destroy the stack and all of its infrastructure with [`pulumi destroy`]({{< relref "/docs/reference/cli/pulumi_destroy" >}}):

```bash
$ pulumi destroy
```

## Learn more

Congratulations! You're now well on your way to managing a production-grade container service on AWS with Pulumi --- and there's lots more you can do from here:

* Discover more architecture templates as they're available in [Templates &rarr;]({{< relref "/templates" >}})
* Dive into the AWS package by exploring the [API docs in the Registry &rarr;]({{< relref "/registry/packages/aws" >}})
* Dive into the AWSx (Crosswalk) package by exploring the [API docs in the Registry &rarr;]({{< relref "/registry/packages/awsx" >}})
* Expand your understanding of how Pulumi works in [Architecture &amp; Concepts &rarr;]({{< relref "/docs/intro/concepts" >}})

> * Read up on the latest new features [in the Pulumi Blog &rarr;]({{< relref "/blog/tag/containers" >}})

Reviewer: My comments on serverless apply here. Don't say as they're available. You should point to the Learn repo instead of docs. Then the Pulumi Blog relref thing doesn't work for tags like the serverless one, so you should fix it like Christian did for that other page. 

I would also merge the lines of AWS and AWSx and just hyperlink the name of the package
